astroquery/mast/core.py

```diff
@@ -199,17 +419,193 @@ def _parse_result(self, responses, verbose=False):
+                          InputWarning)
+            session_token = None
+        elif session_token and store_password:
+            warnings.warn("Password is not used for token based login, therefor password cannot be stored.",
+                          InputWarning)
+
+        if session_token:
+            return self._attach_cookie(session_token)
+        else:
+            # get username if not supplied
+            if not username:
+                username = input("Enter your username: ")
+
+            # check keyring get password if not supplied
+            if not password and not reenter_password:
+                password = keyring.get_password("astroquery:mast.stsci.edu", username)
+
+            # get password if no password is found (or reenter_password is set)
+            if not password:
+                password = getpass("Enter password for {}: ".format(username))
+
+            # store password if desired
+            if store_password:
+                keyring.set_password("astroquery:mast.stsci.edu", username, password)
+
+            return self._shib_login(username, password)
+
+    def login(self, username=None, password=None, session_token=None,
+              store_password=False, reenter_password=False):  # pragma: no cover
+        """
+        Log into the MAST portal.
+
+        Parameters
+        ----------
+        username : string, optional
+            Default is None.
+            The username for the user logging in.
+            Usually this will be the user's email address.
+            If a username is necessary but not supplied it will be prompted for.
+        password : string, optional
+            Default is None.
+            The password associated with the given username.
+            For security passwords should not be typed into the terminal or jupyter
+            notebook, but input using a more secure method such as `~getpass.getpass`.
+            If a password is necessary but not supplied it will be prompted for.
+        session_token : dict or `~http.cookiejar.Cookie`, optional
+            A valid MAST session cookie that will be attached to the current session
+            in lieu of logging in with a username/password.
+            If username and/or password is supplied, this argument will be ignored.
+        store_password : bool, optional
+            Default False.
+            If true, username and password will be stored securely in your keyring.
+        reenter_password : bool, optional
+            Default False.
+            Asks for the password even if it is already stored in the keyring.
+            This is the way to overwrite an already stored password on the keyring.
+        """
+
+        return super(MastClass, self).login(username=username, password=password, session_token=session_token,
+                                            store_password=store_password, reenter_password=reenter_password)
+
+    def logout(self):  # pragma: no cover
```

Comment:
What does `pragma: no cover` mean?  I'm not familiar with the term.

Reply:
It is to exclude that block from coverage calculation. You mark it as such when you know that code is not tested.

Reply:
There is also `# pragma: py2` for tested in Python 2 only and `# pragma: py3` for Python 3 only. Might be others too but I only used these.

Reply:
it just tells coveralls to ignore this (as it was a conscious decision that this would/could not have test coverage)

Reply:
oops, I haven't seen your reply here @pllim, would not have repeated you...

Reply:
No worries!